airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/component/TableSchemaEvolutionClient.kt

```diff
@@ -57,5 +57,156 @@ interface TableSchemaEvolutionClient {
+        columnNameMapping: ColumnNameMapping,
+        tableName: TableName,
+        expectedSchema: TableSchema,
+        expectedAdditionalInfo: AdditionalSchemaInfo,
+        diff: TableSchemaDiff,
+        additionalSchemaInfoDiff: AdditionalSchemaInfoDiff
+    ) {
+        throw NotImplementedError()
+    }
+}
+
+data class TableSchema(
+    /**
+     * A map from column name to type. Note that the column name should be as it appears in the
+     * destination: for example, Snowflake upcases all identifiers, so these column names should be
+     * upcased.
+     */
+    val columns: Map<String, ColumnType>
+) {
+    /** Generate a diff which, when applied to `this`, will result in [expectedSchema]. */
+    fun diff(expectedSchema: TableSchema): TableSchemaDiff {
+        val actualColumns = this.columns
+        val expectedColumns = expectedSchema.columns
+
+        return TableSchemaDiff(
+            columnsToAdd = expectedColumns.filter { !actualColumns.contains(it.key) },
+            columnsToDrop = actualColumns.filter { !expectedColumns.contains(it.key) },
+            columnsToChange =
+                actualColumns
+                    .filter { (name, actualType) ->
+                        expectedColumns.containsKey(name) && expectedColumns[name] != actualType
+                    }
+                    .mapValues { (name, actualType) ->
+                        ColumnTypeChange(
+                            originalType = actualType,
+                            newType = expectedColumns[name]!!,
+                        )
+                    },
+            columnsToRetain =
+                actualColumns.filter { (name, actualType) ->
+                    expectedColumns.containsKey(name) && expectedColumns[name] == actualType
+                },
+        )
+    }
+}
+
+data class ColumnType(
+    /**
+     * A string representation of the data type. For most destinations, this will likely be strings
+     * like "VARCHAR" or "INTEGER".
+     *
+     * Implementations _may_ include precision information (e.g. "VARCHAR(1234)"), but should take
+     * care that their `discoverSchema` and `computeSchema` implementations generate the same
+     * precision.
+     */
+    val type: String,
+    /**
+     * Note that column nullability is not always the same as whether [DestinationStream.schema]
+     * declares a field as "nullable". Many destinations default all user-configured fields to
+     * nullable, and use "nonnull" for other purposes (e.g. Clickhouse and Iceberg require primary
+     * key columns to be non-nullable).
+     */
+    val nullable: Boolean,
+)
+
+/**
+ * As with [TableSchema], all maps are keyed by the column name as it appears in the destination.
+ */
+data class TableSchemaDiff(
+    val columnsToAdd: Map<String, ColumnType>,
+    val columnsToDrop: Map<String, ColumnType>,
+    val columnsToChange: Map<String, ColumnTypeChange>,
+    val columnsToRetain: Map<String, ColumnType>,
+) {
```

Comment:
Our notion of diff here kind of mixes 2 concerns:
1) the actual raw diff between 2 sets
2) how to act on that set

I wonder if it should exclude 2) and be more the difference between 2 sets or the relative complement or what have you. 2) would then be more a function of applyDiff

Reply:
For posterity, I think we're good to stick with this approach but call it `ChangeSet` instead of diff for clarity.